src/dataflow/src/sink/tail.rs

```diff
@@ -94,14 +98,68 @@ pub fn tail<G>(
                 return;
             }
         });
+
+        let progress_row = update_progress(
+            &mut input_frontier,
+            input.frontier().frontier(),
+            &mut packer,
+            connector.object_columns + 1,
+        );
+
+        // Only emit updates if this operator/worker received actual
+        // data for emission. For TAIL, data is exchanged to one worker,
+        // which forwards all the data to the client process. If we
+        // blindly forwarded the frontier from all workers we would get
+        // multiple progress updates in the client.
+        if connector.emit_progress && received_data && progress_row.is_some() {
+            let results = vec![progress_row.expect("known to exist")];
+
+            // TODO(benesch): the lack of backpressure here can result in
+            // unbounded memory usage.
+            if connector.tx.send(results).is_err() {
+                errored = true;
+                return;
+            }
+        }
     })
 }
 
-fn batch_upper(
-    batch: Option<&Rc<OrdValBatch<GlobalId, Row, u64, isize, usize>>>,
-) -> Option<Timestamp> {
-    batch
-        .map(|b| b.desc.upper().elements().get(0))
-        .flatten()
-        .copied()
+// Checks if there is progress between `current_input_frontier` and
+// `new_input_frontier`. If yes, updates the `current_input_frontier` to
+// `new_input_frontier` and returns a [`Row`] that encodes this progress. If
+// there is no progress, returns [`None`].
+fn update_progress(
+    current_input_frontier: &mut Vec<Timestamp>,
+    new_input_frontier: AntichainRef<Timestamp>,
+    packer: &mut Row,
+    empty_columns: usize,
+) -> Option<Row> {
+    // Test to see if strict progress has occurred (any of the old frontier less equal
+    // to the new frontier).
+    let progress = current_input_frontier
+        .iter()
+        .any(|t2| !new_input_frontier.less_equal(t2));
+
+    if progress {
+        current_input_frontier.clear();
+        current_input_frontier.extend(new_input_frontier.iter().cloned());
+
+        // This only looks at the first entry of the antichain.
+        // If we ever have multi-dimensional time, this is not correct
+        // anymore. There might not even be progress in the first dimension.
+        let upper = new_input_frontier.first().cloned();
```

Comment:
We can use `Itertools::exactly_one` here to make sure this produces a panic if we ever have more than one thing in the frontier

Reply:
`exactly_once` would not be correct, because this can be empty. And empty Antichain signals that there will be no more input, which happens when the source shuts down.

I could add an assert to check that there are not more than one entries, though.

Reply:
then [Itertools::at_most_one](https://docs.rs/itertools/0.10.1/itertools/trait.Itertools.html#method.at_most_one)

Reply:
Thanks, you truly do know your rust! :sweat_smile: